File: scripts/observations/Lcom4Collector.php
<?php

declare(strict_types=1);

/**
 * LCOM4 observation collector (Hitz & Montazeri).
 *
 * LCOM4 = number of connected components in the method graph of a class,
 * where methods are connected when they share an instance variable or one
 * uses the other as a behaviour of the same class. Constructors/destructors
 * are excluded (they touch everything and would mask real cohesion splits).
 *
 * The internal behavioural relationship is determined by the dependency, NOT
 * by the syntax expressing it: $this->m(), self::m(), static::m() and
 * OwnClass::m() are the same relationship. parent::m() is not — it uses
 * inherited behaviour outside the analysed class. Dynamic targets and
 * first-class callables are outside the contract. The pinned decisions in
 * examples/lcom4/expected.json are the source of truth; this file conforms
 * to them, never the reverse.
 *
 * Scope guard: emits OBSERVATIONS only — metric · class · value ·
 * interpretation. No thresholds, no warnings, no verdicts. Whether a value
 * matters is configured governance's question, never the collector's.
 *
 * Parsing: nikic/php-parser (already a repo dependency — reused, not built).
 */

require_once __DIR__ . '/../../vendor/autoload.php';

use PhpParser\Node;
use PhpParser\NodeFinder;
use PhpParser\ParserFactory;

final class Lcom4Collector
{
    /**
     * @return list<array{metric:string,class:string,value:int,interpretation:string}>
     */
    public static function collect(string $phpSource): array
    {
        $parser = (new ParserFactory())->createForNewestSupportedVersion();
        $ast = $parser->parse($phpSource) ?? [];
        $finder = new NodeFinder();

        $observations = [];
        foreach ($finder->findInstanceOf($ast, Node\Stmt\Class_::class) as $class) {
            $observations[] = self::observeClass($class, $finder);
        }
        return $observations;
    }

    /** @return array{metric:string,class:string,value:int,interpretation:string} */
    private static function observeClass(Node\Stmt\Class_ $class, NodeFinder $finder): array
    {
        $ownName = $class->name?->toString();

        $methods = [];   // name => ['props' => set, 'calls' => set]
        foreach ($class->getMethods() as $method) {
            $name = $method->name->toString();
            if (in_array($name, ['__construct', '__destruct'], true)) {
                continue;
            }
            $props = [];
            $calls = [];
            foreach ($finder->findInstanceOf($method, Node\Expr\PropertyFetch::class) as $fetch) {
                if ($fetch->var instanceof Node\Expr\Variable && $fetch->var->name === 'this'
                    && $fetch->name instanceof Node\Identifier) {
                    $props[$fetch->name->toString()] = true;
                }
            }
            // Internal behavioural relationship, reached through the object.
            foreach ($finder->findInstanceOf($method, Node\Expr\MethodCall::class) as $call) {
                if ($call->var instanceof Node\Expr\Variable && $call->var->name === 'this'
                    && $call->name instanceof Node\Identifier
                    && !self::isFirstClassCallable($call)) {
                    $calls[$call->name->toString()] = true;
                }
            }
            // Same relationship, reached through the class (self:: / static:: /
            // OwnClass::). parent:: is deliberately NOT here — see intra_class_calls.
            foreach ($finder->findInstanceOf($method, Node\Expr\StaticCall::class) as $call) {
                if ($call->class instanceof Node\Name
                    && $call->name instanceof Node\Identifier
                    && self::namesThisClass($call->class, $ownName)
                    && !self::isFirstClassCallable($call)) {
                    $calls[$call->name->toString()] = true;
                }
            }
            $methods[$name] = ['props' => array_keys($props), 'calls' => array_keys($calls)];
        }

        $value = self::connectedComponents($methods);

        return [
            'metric'         => 'LCOM4',
            'class'          => $class->name !== null ? $class->name->toString() : '(anonymous)',
            'value'          => $value,
            'interpretation' => match (true) {
                $value === 0 => 'no analyzable methods',
                $value === 1 => 'single connected component (cohesive)',
                default      => sprintf('%d connected components (disjoint responsibility clusters)', $value),
            },
        ];
    }

    /**
     * Does this static call name a method OF THIS CLASS?
     *
     * self:: and static:: do by definition. The class's own name spelled out is
     * the same relationship written differently. parent:: does NOT — it targets
     * inherited behaviour, which is outside the analysed class (inherited_methods).
     */
    private static function namesThisClass(Node\Name $class, ?string $ownName): bool
    {
        $ref = $class->toString();
        if (in_array(strtolower($ref), ['self', 'static'], true)) {
            return true;
        }
        if (strtolower($ref) === 'parent') {
            return false;
        }
        // Own name as written. Namespaced/aliased spellings are not resolved —
        // the class is analysed in isolation (known limitation).
        return $ownName !== null && strtolower($ref) === strtolower($ownName);
    }

    /**
     * First-class callable syntax — m(...) — REFERENCES a behaviour rather than
     * invoking it, and is outside this contract by decision (PO/ARB 2026-08-16).
     */
    private static function isFirstClassCallable(Node\Expr\MethodCall|Node\Expr\StaticCall $call): bool
    {
        foreach ($call->args as $arg) {
            if ($arg instanceof Node\VariadicPlaceholder) {
                return true;
            }
        }
        return false;
    }

    /** @param array<string,array{props:list<string>,calls:list<string>}> $methods */
    private static function connectedComponents(array $methods): int
    {
        $names = array_keys($methods);
        if ($names === []) {
            return 0;
        }

        // union-find over method names
        $parent = array_combine($names, $names);
        $find = function (string $m) use (&$parent, &$find): string {
            return $parent[$m] === $m ? $m : $parent[$m] = $find($parent[$m]);
        };
        $union = function (string $a, string $b) use (&$parent, $find): void {
            $parent[$find($a)] = $find($b);
        };

        // edge: shared instance variable
        $propOwners = [];
        foreach ($methods as $name => $info) {
            foreach ($info['props'] as $prop) {
                if (isset($propOwners[$prop])) {
                    $union($name, $propOwners[$prop]);
                } else {
                    $propOwners[$prop] = $name;
                }
            }
        }
        // edge: method call within the class
        foreach ($methods as $name => $info) {
            foreach ($info['calls'] as $callee) {
                if (isset($methods[$callee])) {
                    $union($name, $callee);
                }
            }
        }

        return count(array_unique(array_map($find, $names)));
    }
}

File: scripts/observations/lcom4_collector.py
#!/usr/bin/env python3
"""LCOM4 observation collector — independent Python implementation.

WHY THIS EXISTS
    Stage 2 of KOS-CONTRACT-NEUTRALITY-001. It answers one question:
    can the corrected language-neutral LCOM4 contract be implemented
    independently and produce the same observations as the PHP reference?

    It is an EVIDENCE EXPERIMENT. It adopts no language, opens no gate, and
    decides nothing about whether PHP is retained or retired.

WHAT IT IS IMPLEMENTED FROM
    scripts/observations/examples/lcom4/expected.json — the pinned decisions
    are the specification. This is deliberately NOT a port of
    Lcom4Collector.php: porting would test transliteration rather than whether
    the written contract is sufficient, which is the entire question.

    The parsing strategy is intentionally different from the reference's. The
    reference uses a full PHP AST parser (nikic/php-parser). No such parser is
    available here and installing one is forbidden by the grant, so this reads
    the source with its own scanner. If two such different strategies agree,
    the agreement is about the CONTRACT, not about a shared library.

CONTRACT AS IMPLEMENTED (each line traceable to a pinned decision)
    LCOM4            = number of connected components over the class's methods
    nodes            = methods declared in the class body
                       - __construct/__destruct excluded          [constructors]
                       - statics included                       [static_methods]
                       - trait methods not resolved              [trait_methods]
                       - inherited not included             [inherited_methods]
                       - no other magic-method handling          [magic_methods]
    edge, state      = two methods touch the same $this->property
    edge, behaviour  = one method uses another as a behaviour of the same
                       class:  $this->m()  self::m()  static::m()  Own::m()
                                                            [intra_class_calls]
    NOT an edge      = parent::m()            (out of frame — nothing to link)
                     = $this->$n(), call_user_func([...]), $var::m()
                                              (not determinable — cannot see it)
                     = m(...)                          [first_class_callables]

    The two exclusions above are different claims and are kept apart in the
    code as they are in the contract: OUT_OF_FRAME vs NOT_DETERMINABLE.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

# --------------------------------------------------------------------------
# 1 · Scanning: blank out comments and string literals so that later pattern
#     matching cannot be fooled by text inside them. Lengths are preserved so
#     every offset stays valid.
# --------------------------------------------------------------------------


def blank_noise(src: str) -> str:
    out = list(src)
    i, n = 0, len(src)
    while i < n:
        c = src[i]
        if c == "/" and i + 1 < n and src[i + 1] == "/":
            j = src.find("\n", i)
            j = n if j == -1 else j
            for k in range(i, j):
                out[k] = " "
            i = j
        elif c == "#":
            j = src.find("\n", i)
            j = n if j == -1 else j
            for k in range(i, j):
                out[k] = " "
            i = j
        elif c == "/" and i + 1 < n and src[i + 1] == "*":
            j = src.find("*/", i + 2)
            j = n if j == -1 else j + 2
            for k in range(i, j):
                if src[k] != "\n":
                    out[k] = " "
            i = j
        elif c in "'\"":
            quote, j = c, i + 1
            while j < n:
                if src[j] == "\\":
                    j += 2
                    continue
                if src[j] == quote:
                    j += 1
                    break
                j += 1
            for k in range(i + 1, min(j - 1, n) + 1):
                if k < n and src[k] != "\n":
                    out[k] = " "
            i = j
        else:
            i += 1
    return "".join(out)


def match_block(src: str, open_pos: int) -> int:
    """Index just past the '}' closing the '{' at open_pos."""
    depth, i, n = 0, open_pos, len(src)
    while i < n:
        if src[i] == "{":
            depth += 1
        elif src[i] == "}":
            depth -= 1
            if depth == 0:
                return i + 1
        i += 1
    return n


# --------------------------------------------------------------------------
# 2 · Structure: classes, then the methods declared in each class body.
#     Only `class` — a trait/interface/enum is not a class    [trait_methods]
# --------------------------------------------------------------------------

CLASS_RE = re.compile(r"\bclass\s+([A-Za-z_]\w*)", re.I)
METHOD_RE = re.compile(r"\bfunction\s+([A-Za-z_]\w*)\s*\(", re.I)
EXCLUDED_METHODS = {"__construct", "__destruct"}  # [constructors]


def find_classes(clean: str):
    for m in CLASS_RE.finditer(clean):
        brace = clean.find("{", m.end())
        if brace == -1:
            continue
        yield m.group(1), brace, match_block(clean, brace)


def find_methods(clean: str, body_start: int, body_end: int):
    """Methods declared directly in the class body (depth 1 — not in closures)."""
    for m in METHOD_RE.finditer(clean, body_start, body_end):
        # depth relative to the class body: 1 means "directly in the class"
        depth = clean.count("{", body_start, m.start()) - clean.count(
            "}", body_start, m.start()
        )
        if depth != 1:
            continue
        brace = clean.find("{", m.end())
        semi = clean.find(";", m.end())
        if brace == -1 or (semi != -1 and semi < brace):
            yield m.group(1), None, None  # abstract/interface: declared, no body
            continue
        yield m.group(1), brace, match_block(clean, brace)


# --------------------------------------------------------------------------
# 3 · Relationships inside one method body.
# --------------------------------------------------------------------------

PROP_OR_CALL_RE = re.compile(r"\$this\s*->\s*([A-Za-z_]\w*)")
STATIC_CALL_RE = re.compile(r"(\$?[A-Za-z_\\]\w*)\s*::\s*([A-Za-z_]\w*)\s*\(")
FIRST_CLASS_CALLABLE_RE = re.compile(r"\A\s*\.\.\.\s*\)")  # m(...)  [first_class_callables]

OUT_OF_FRAME = {"parent"}          # nothing to connect to
SELF_REFERENTIAL = {"self", "static"}


def analyse_method(body: str, own_class: str):
    """-> (properties touched, methods used as behaviours of this class)"""
    props: set[str] = set()
    calls: set[str] = set()

    for m in PROP_OR_CALL_RE.finditer(body):
        rest = body[m.end():]
        stripped = rest.lstrip()
        if not stripped.startswith("("):
            props.add(m.group(1))          # $this->state — shared state edge
            continue
        after_paren = stripped[1:]
        if FIRST_CLASS_CALLABLE_RE.match(after_paren):
            continue                        # $this->m(...) — reference, excluded
        calls.add(m.group(1))               # $this->m()  — behavioural edge

    for m in STATIC_CALL_RE.finditer(body):
        qualifier, method = m.group(1), m.group(2)
        if qualifier.startswith("$"):
            continue                        # $var::m() — NOT DETERMINABLE
        q = qualifier.lower()
        if q in OUT_OF_FRAME:
            continue                        # parent::m() — OUT OF FRAME
        own_as_written = q == own_class.lower()
        if not (q in SELF_REFERENTIAL or own_as_written):
            continue                        # another class entirely
        if FIRST_CLASS_CALLABLE_RE.match(body[m.end():]):
            continue                        # self::m(...) — reference, excluded
        calls.add(method)

    return props, calls


# --------------------------------------------------------------------------
# 4 · Connected components (union-find over method names).
# --------------------------------------------------------------------------


def connected_components(methods: dict[str, tuple[set, set]]) -> int:
    names = list(methods)
    if not names:
        return 0
    parent = {n: n for n in names}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    owner: dict[str, str] = {}
    for name, (props, _) in methods.items():
        for p in props:
            if p in owner:
                union(name, owner[p])
            else:
                owner[p] = name

    for name, (_, calls) in methods.items():
        for callee in calls:
            if callee in methods:          # only methods declared in this class
                union(name, callee)

    return len({find(n) for n in names})


# --------------------------------------------------------------------------
# 5 · Observation
# --------------------------------------------------------------------------


def interpretation(value: int) -> str:
    if value == 0:
        return "no analyzable methods"
    if value == 1:
        return "single connected component (cohesive)"
    return f"{value} connected components (disjoint responsibility clusters)"


def collect(php_source: str) -> list[dict]:
    clean = blank_noise(php_source)
    observations = []
    for class_name, body_start, body_end in find_classes(clean):
        methods: dict[str, tuple[set, set]] = {}
        for name, m_start, m_end in find_methods(clean, body_start, body_end):
            if name in EXCLUDED_METHODS:
                continue
            body = "" if m_start is None else clean[m_start:m_end]
            methods[name] = analyse_method(body, class_name)
        value = connected_components(methods)
        observations.append(
            {
                "metric": "LCOM4",
                "class": class_name,
                "value": value,
                "interpretation": interpretation(value),
            }
        )
    return observations


def main() -> int:
    if len(sys.argv) > 1:
        for path in sys.argv[1:]:
            print(json.dumps(collect(Path(path).read_text()), indent=2))
        return 0
    print(json.dumps(collect(sys.stdin.read()), indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
